Add callback with TLS state API in futures

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

rohan-varma committed:
Add callback with TLS state API in futures

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -630,6 +631,17 @@ getClassConverter() {
+          at::ThreadLocalStateGuard g(tls_state);
+          callback();
+        };
+    addCallback(cbWithTlsState);
```
Would I be correct if I assume this will result in a copy (including `tls_state` and `callback`)?

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -338,19 +338,30 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   }
 
+  /**
+  * Add a callback to the future that that runs with the thread local state of
```
that that -> that

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -338,19 +338,30 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   c10::intrusive_ptr<Future> then(
       std::function<IValue(void)> callback,
-      TypePtr type) {
+      TypePtr type, bool propagateTLSState = false) {
```
break into two lines?

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -338,19 +338,30 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     auto fut = c10::make_intrusive<Future>(type);
+    // Pick callback function based on propagateTLSState
+    void (ivalue::Future::*addCallbackFn)(std::function<void(void)>) =
+        propagateTLSState ? &ivalue::Future::addCallbackWithTLSState
```
can `propagateTLSState` be a argument in `Future::addCallback` instead of having two methods + manipulating function pointers?

rohan-varma replied:
Sure, I think that this is reasonable as well.

mrshenli commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2048,6 +2048,52 @@ void testFutures() {
+  std::mutex cbFinishedMutex;
+  std::condition_variable cbFinishedCv;
+  // cb that verifies the profiler is enabled
+    auto profilerEnabledCb = [&cbFinished, &cbFinishedMutex, &cbFinishedCv](){
```
indent

jjlilley commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2048,6 +2048,52 @@ void testFutures() {
     ASSERT_EQ(c4->value().toInt(), 7);
   }
+
+  bool cbFinished = false;
```
maybe it's worth separating these TLSState-oriented tests in a separate test function?
(add the test function in caffe2/test/cpp/jit/tests.h)

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallbackWithTLSState(some_cb_that_requires_tls_state);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(some_cb_that_requires_tls_state, true);
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma commented:
ping @mrshenli @jjlilley

jjlilley commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -630,6 +631,29 @@ getClassConverter() {
+    std::function<void(void)> callback,
+    bool propagateTLSState) {
+  std::function<void(void)> cb;
+  if (propagateTLSState) {
```
I'm fine with this approach as-is. But before we land this, I wanted to throw out an alternative to see if we considered it.

In the spirit of unix-y composition of small components, instead of
  `future->addCallback(cb, true /*propagateTLSState*/);`
add optional wrapper like the following:
  `future->addCallback(wrapPropagateTLSState(cb));`

The impl would be basically cutting and pasting the bits we already have, and wouldn't need to touch addCallback(), anymore, just add a short utility function like:

  std::function<void(void)> wrapPropagateTLSState(std::function<void(void)> cb) {
    at::ThreadLocalState tls_state;
    return [tls_state = std::move(tls_state), callback = std::move(callback)] {
        at::ThreadLocalStateGuard g(tls_state);
        callback();
      };
  }

The advantage would mostly be a simpler base impl with no implicit args, and perhaps more maintainability if we ever want to add another arg to the `addCallback` function (hopefully that could be done via composition as well)

rohan-varma replied:
This sounds good to me, as we won't have to modify addCallback and then() at all.

pritamdamania87 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -630,6 +631,29 @@ getClassConverter() {
+  if (propagateTLSState) {
+    // Save thread local state and restore it in the callback.
+    at::ThreadLocalState tls_state;
+    cb = [tls_state = std::move(tls_state), callback = std::move(callback)] {
```
nit: `tls_sate = ThreadLocalState()`

pritamdamania87 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -630,6 +631,29 @@ getClassConverter() {
 
+void ivalue::Future::addCallback(
+    std::function<void(void)> callback,
+    bool propagateTLSState) {
```
Do we need this flag? Why don't we always propagate the TLS state? (similar to what we do in at::launch()). If this is optional, we'll mostly run in to issues time and again where we encounter a bug and realize we need to set this to true.

If we would end up breaking some existing code with such a change, maybe make the default value of the flag true and fix the handful of places by setting this to false.

jjlilley replied:
ivalue::Future seems to be a fairly generic-intended api.

It feels like a strong position to say that we unconditionally want it to propagate TLS state - this will get expensive over time, and I can imagine low-level operations where this might be be particularly appropriate.

Supporting it here with a wrapper or param seems reasonable, to avoid boilerplate.
But forcing it in every case may be excessively heavyweight.

rohan-varma replied:
I think propagating it by default might warrant some more discussion. Maybe for this PR we can add @jjlilley 's suggestion of the wrapper function, and use in RPC-related callsites. If we do decide to unconditionally propagate it should be an easy change to just have `addCallback` call into this helper, and remove the helper from all callsites.

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -345,7 +339,8 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
       std::function<IValue(void)> callback,
-      TypePtr type) {
+      TypePtr type,
+      bool propagateTLSState = false) {
```
Same as the comment in `addCallback`, would prefer not having this flag at all and make propagation the default.

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -630,6 +631,17 @@ getClassConverter() {
   return classConverter;
 }
 
+template <typename T>
```
I'm not sure if this is the best way to approach this, but the reason this function needs to be templated is because it needs to propagate the value of `callback()`. Normally the template would be defined in the header to avoid the explicit instantiation below, but there were some compile errors when adding the necessary `include <ATen/ThreadLocalState.h>` to `ivalue_inl.h` (it looks like it caused a circular header issue).

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

jjlilley reviewed (commented):
fwiw, the approach/code here LGTM, though I'll wait to approve in case others see gotchas

jjlilley commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -119,38 +119,31 @@ using namespace torch::distributed::autograd;
-          at::ThreadLocalStateGuard g(tls_state);
-          if (futureResponseMessage.hasError()) {
-            jitFuture->setError(futureResponseMessage.error()->what());
+        at::wrapPropagateTLSState<void>([jitFuture, futureResponseMessage]() {
```
Actually - are we sure that this case (and some others) don't lead to a memory leak?

The ones that take "const FutureMessage&" arg - I think these were actually mostly on purpose. When I was refactoring the futures months ago, my initial check-in caused a memory regression, e.g. in ctr_mobile_feed.  =(

And it took a bit of effort to figure out which was causing the regression. IIRC, sometimes the callbacks weren't being [ever] called. Without the ref cycle, things would be ok - they'd just vanish. But when there's a reference cycle like this, they could never be freed, and we'd leak memory pretty quickly.

My suggestion here would be maybe take an explicit weak reference for futureResponseMesssage. Assert that the result is non-null when you promote it to strong shared_ptr. Then everything should be ok with the wrapping.

But if you take a strong shared_ptr reference in these cases, there might be some unfortunate side effects.

rohan-varma replied:
I see, the reason that I refactored out the `const FutureMessage& futureResponseMessage` argument types was because the `wrapPropagateTLSState` API only accepts functions which take no args, and I thought that the `addCallback` API here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/future.h#L135 was mostly a helper function that this code is equivalent to. It looks like it's not equivalent though, because in the case where the callback isn't called we would never deallocate `futureResponseMessage`? 

In that function, it looks like we make a copy of the `this` pointer?

jjlilley replied:
Right, I understand the motivation with the new function wrapper - keeping without arguments is good, but we just have to avoid capturing a shared_ptr in the lambda captures.

The helper `const FutureMessage&` felt slightly hacky (though is safe in its uses) - it's effectively just passing in an unsafe pointer to avoid the reference cycle. (The reference cycle causing `futureResponseMessage` never to be dealloced, and a slow leak)

You could pass in the unsafe pointer into the lambda capture, though it might be better-received than if you captured the weak reference instead:
```
  std:weak_ptr<FutureMessage> wp = futureResponseMessage;
  addCallback([wp]() {
    auto futureResponseMessage = wp.lock();
    ...
  });
```

rohan-varma committed:
Update on "Add callback with TLS state API in futures"

enum.
* **#40326 Add callback with TLS state API in futures**
enum.
* **#40326 Add callback with TLS state API in futures**

Adds a helper function `addCallbackWithTLSState` to both
torch/csrc/utils/future.h which is used internally by RPC framework and the JIT
future. Uses this helper function to avoid to pass in TLS state where it is needed for rpc and `record_function_ops.cpp`. For example, the following:

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -119,38 +119,35 @@ using namespace torch::distributed::autograd;
   if (hasValue) {
     c10::intrusive_ptr<JitFuture> jitFuture =
         c10::make_intrusive<JitFuture>(PyObjectType::get());
+    std::weak_ptr wp = futureResponseMessage;
```
I think you need the templated `weak_ptr` type (e.g. `std::weak_ptr<FutureMessage>`) to appease the compiler

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

jjlilley reviewed (approved):
Looks great to me, thanks!

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add callback with TLS state API in futures"


Adds a helper function `wrapPropagateTLSState` that wraps a function into another function that runs with the `at::ThreadLocalState` of the caller. The motivation is to compose a function with this wrapper when calling APIs such as `addCallback` or `then` so that different threads run callbacks with the right tls state.

With the API, the below

```
at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
at::ThreadLocalStateGuard g(tls_state);
some_cb_that_requires_tls_state();
}
```

becomes

```
fut->addCallback(wrapPropagateTLSState<returnType>(some_cb_that_requires_state));
```

Differential Revision: [D22147634](https://our.internmc.facebook.com/intern/diff/D22147634/)

[ghstack-poisoned]

rohan-varma commented:
Had to move the helper to `ATen/ThreadLocalState.h` since there were a bunch of build issues internally.
